### PR TITLE
Genotype concordance workflows

### PR DESCRIPTION
**Pull request by mwalker174**
Adds two new WDLs:

1. JoinRawCalls - merges raw variant calls across batches from `ClusterBatch` outputs
2. SVConcordance - evaluates genotype concordance of an "eval" vcf against a "truth" vcf.

The second workflow is intended to be run with a "cleaned" eval vcf and a joined raw call truth vcf from the same cohort. The output vcf can then be used for filtering with `TrainGqRecalibrator` and `RecalibrateGq` workflows as the `annotations_vcf` input, along with the appropriate `annotations_to_transfer` list (TBD).

Note this depends on a [pending PR in gatk](https://github.com/broadinstitute/gatk/pull/7977), so I've added a temporary concordance gatk docker.

Also includes necessary updates to the vcf formatting script. Currently the concordance tool does not handle CPX/CTX so those are filtered out. As a result, the concordance-annotated VCF should NOT be carried forward other than as an input to the GQ recalibrator workflows.

Tested and updated both new workflows with inputs generated from corresponding templates.

### CODE REVIEW

**Review by TedBrookings — CHANGES_REQUESTED** [review]
I have a few questions and I'm requesting a few minor, stylistic changes. Overall it looks pretty good.

I have a couple comments about format_svtk_vcf_for_gatk.py that are either 1) not really a request for change or 2) just outside the changes you made and therefore not critical for the PR
1) Minor style note: I noticed __parse_bnd_ends and __parse_ploidy_table start with double-underscores. I've taken to using double-underscores for __parse_arguments because double-underscores enforce privacy. That way, if a python script does "from other_script import *" and that other script happens to have __parse_arguments in it, nothing bad happens, but if they both had _parse_arguments, that would get clobbered and produce crazy crashes. I don't think it's necessary to use more than single underscore for most other private functions, and it's arguably overkill even for __parse_arguments.

2) It's better practice to use a context manager than directly handling the closing of file-like objects. e.g.
```
with vcf_in = pysam.VariantFile(arguments.vcf, 'r')
    # vcf_in will be closed after block ends, including early return, exception, etc
    do_stuff()
```
Technically you could do that even with filter_out, using nullcontext, although that may be a bit clunky.

File: inputs/templates/test/SVConcordance/SVConcordance.json.tmpl [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  "SVConcordance.ploidy_table": {{ test_batch.ploidy_table | tojson }},
+  "SVConcordance.cohort": {{ test_batch.name | tojson }},
+
+  "SVConcordance.preprocess_truth_vcf": "false",
+  "SVConcordance.preprocess_eval_vcf": "true",
+  "SVConcordance.preprocess_eval_args": "--only-add-cn-fields --replace-ev-format --filter-unsupported-types",
+
+  "SVConcordance.contig_list": {{ reference_resources.primary_contigs_list | tojson }},
+  "SVConcordance.reference_dict": {{ reference_resources.reference_dict | tojson }}
+}
```
Comment (TedBrookings) [inline inputs/templates/test/SVConcordance/SVConcordance.json.tmpl:19]: Add a newline
```suggestion
}

```

File: src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py [hunk 2]
```diff
@@ -222,6 +299,16 @@ def __parse_arguments(argv: List[Text]) -> argparse.Namespace:
                              "first column is SAMPLE, and the remaining columns are contig names. For each row "
                              "thereafter, the first column is the sample name, and remaining columns are the contig "
                              "ploidy values for that sample.")
+    parser.add_argument("--only-add-cn-fields", action='store_true',
+                        help="Only add CN and ECN info fields. All other corrections are skipped.")
+    parser.add_argument("--use-end2", action='store_true',
+                        help="Use existing END2 fields rather than getting them from END")
+    parser.add_argument("--filter-unsupported-types", action='store_true',
+                        help="Filter CPX and CTX types, which are not currently supported by GATK")
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py:307]: Agreed, I made a ticket [here](https://github.com/broadinstitute/gatk/issues/8033)

File: src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py [hunk 3]
```diff
@@ -222,6 +299,16 @@ def __parse_arguments(argv: List[Text]) -> argparse.Namespace:
                              "first column is SAMPLE, and the remaining columns are contig names. For each row "
                              "thereafter, the first column is the sample name, and remaining columns are the contig "
                              "ploidy values for that sample.")
+    parser.add_argument("--only-add-cn-fields", action='store_true',
+                        help="Only add CN and ECN info fields. All other corrections are skipped.")
+    parser.add_argument("--use-end2", action='store_true',
+                        help="Use existing END2 fields rather than getting them from END")
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py:305]: I don't think I've encountered this case. I'm hesitant to change it now, but if we do come across a mix we can fix later.

File: src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py [hunk 4]
```diff
@@ -200,6 +231,52 @@ def convert(record: pysam.VariantRecord,
     return new_record
 
 
+def _cache_gt_sum(gt):
+    if gt in _gt_sum_map:
+        return _gt_sum_map[gt]
+    s = sum([1 for a in gt if a is not None and a > 0])
+    _gt_sum_map[gt] = s
+    return s
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py:239]: I usually forget about the `get()` method which is very handy. I've made the suggested change.

File: src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py [hunk 5]
```diff
@@ -200,6 +231,52 @@ def convert(record: pysam.VariantRecord,
...
+
+
+def add_cn_ecn(record: pysam.VariantRecord,
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py:242]: Done

File: src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py [hunk 6]
```diff
@@ -200,6 +231,52 @@ def convert(record: pysam.VariantRecord,
...
+               vcf_out: pysam.VariantFile,
+               ploidy_dict: Dict[Text, Dict[Text, int]]) -> pysam.VariantRecord:
+    svtype = record.info['SVTYPE']
+    contig = record.contig
+    new_record = vcf_out.new_record(contig=contig, start=record.start, stop=record.stop, alleles=record.alleles)
+    new_record.id = record.id
+
+    # copy INFO fields
+    for key in record.info:
+        new_record.info[key] = record.info[key]
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py:252]: Done

File: src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py [hunk 7]
```diff
@@ -96,16 +105,20 @@ def create_header(header_in: pysam.VariantHeader,
         header.add_line(line_str)
     # new fields
     header.add_line('##INFO=<ID=END2,Number=1,Type=Integer,Description="Second breakend position">')
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py:107]: I believe the line will just be replaced if it is

File: wdl/SVConcordance.wdl [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+  input {
+    File eval_vcf
+    File truth_vcf
+
+    File ploidy_table
+    String cohort
+
+    Boolean? preprocess_truth_vcf
+    String? preprocess_truth_args
+    Boolean? preprocess_eval_vcf
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply wdl/SVConcordance.wdl:16]: This causes problems with miniwdl, I believe

File: wdl/SVConcordance.wdl [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    # Convert format
+    python ~{default="/opt/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py" script} \
+      --vcf ~{vcf} \
+      --out tmp.vcf.gz \
+      --filter-out ~{output_prefix}.filtered_records.vcf.gz \
+      --ploidy-table ~{ploidy_table} \
+      ~{args}
+
+    # TODO Filter invalid records with SVLEN=0, only needed for legacy runs that used svtk cluster in ClusterBatch
+    bcftools view --no-version -i 'INFO/SVLEN="." || INFO/SVLEN>0' tmp.vcf.gz -Oz -o ~{output_prefix}.vcf.gz
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply wdl/SVConcordance.wdl:194]: These should be dropped during clustering since they're less than the minimum size of 50 - I'm not sure why they aren't.

File: inputs/values/ref_panel_1kg.json [hunk 10]
```diff
@@ -2236,6 +2238,7 @@
     "name": "ref_panel_1kg",
     "outlier_cutoff_table": "gs://gatk-sv-resources-public/hg38/v0/sv-resources/ref-panel/1KG/v1/module03_outlier_cutoff_table.tsv",
     "ped_file": "gs://gcp-public-data--broad-references/hg38/v0/sv-resources/ref-panel/1KG/v1/ped/1kg_ref_panel_v1.ped",
+    "ploidy_table": "gs://gatk-sv-ref-panel-1kg/outputs/ClusterBatch/mw-concordance/ref_panel_1kg.ploidy.FEMALE_chrY_1.tsv",
```
Comment (epiercehoffman) [inline inputs/values/ref_panel_1kg.json:2241]: Should open a ticket to track changes needed to be able to repopulate this JSON with create_test_batch.py including outputting this file from the Batch pipeline, or keeping this value along with the new batch's metadata. I think some other recent changes might need to sync up with the create_test_batch.py script too (ex. handling tarballs of std vcfs vs. the file lists used in the single-sample pipeline) so it would be good to keep track of those

File: inputs/templates/test/SVConcordance/SVConcordance.json.tmpl [hunk 11]
```diff
@@ -0,0 +1,19 @@
+{
+  "SVConcordance.gatk_docker":{{ dockers.gatk_docker_concordance | tojson }},
+  "SVConcordance.sv_base_mini_docker": {{ dockers.sv_base_mini_docker | tojson }},
+  "SVConcordance.sv_pipeline_docker": {{ dockers.sv_pipeline_docker | tojson }},
+  "SVConcordance.sv_utils_docker": {{ dockers.sv_utils_docker | tojson }},
+
+  "SVConcordance.eval_vcf" : {{ test_batch.merged_depth_vcf | tojson }},
```
Comment (epiercehoffman) [inline inputs/templates/test/SVConcordance/SVConcordance.json.tmpl:7]: I thought the comparison would be the raw calls vs. the clean VCF, not the clustered depth VCF? (If that's the case - the clean VCF should be GATK-ready except for dropping CTX/CPX for your tool, so I'm not sure that the call to SvUtilsFixVcf is necessary)

File: inputs/values/ref_panel_1kg.json (hunk 10 above)
Comment (mwalker174) [reply inputs/values/ref_panel_1kg.json:2244]: Done

File: inputs/templates/test/SVConcordance/SVConcordance.json.tmpl (hunk 11 above)
Comment (mwalker174) [reply inputs/templates/test/SVConcordance/SVConcordance.json.tmpl]: Good catch, fixed

File: src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py [hunk 12]
```diff
@@ -132,30 +146,46 @@ def convert(record: pysam.VariantRecord,
...
     else:
         alleles = record.alleles
     contig = record.contig
-    new_record = vcf_out.new_record(contig=contig, start=record.start, stop=record.stop, alleles=alleles)
-    new_record.id = record.id
+    new_record = vcf_out.new_record(id=record.id, contig=contig, start=record.start, stop=record.stop, alleles=alleles)
     # copy INFO fields
     for key in record.info:
         if key not in default_remove_infos and key not in remove_infos:
             new_record.info[key] = record.info[key]
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/sv-pipeline/scripts/format_svtk_vcf_for_gatk.py]: Made this change

File: inputs/values/dockers.json [hunk 13]
```diff
@@ -28,8 +29,8 @@
   "cnmops-virtual-env": "us.gcr.io/broad-dsde-methods/gatk-sv/cnmops-virtual-env:2022-06-10-v0.23-beta-9c6fbf56",
   "sv-pipeline-virtual-env": "us.gcr.io/broad-dsde-methods/gatk-sv/sv-pipeline-virtual-env:2022-06-10-v0.23-beta-9c6fbf56",
   "samtools-cloud-virtual-env": "us.gcr.io/broad-dsde-methods/gatk-sv/samtools-cloud-virtual-env:2022-06-10-v0.23-beta-9c6fbf56",
-  "sv-utils-env": "us.gcr.io/broad-dsde-methods/gatk-sv/sv-utils-env:2022-09-09-v0.25.1-beta-cd5ce7e0",
-  "sv_utils_docker": "us.gcr.io/broad-dsde-methods/gatk-sv/sv-utils:2022-09-21-v0.26.1-beta-c5e6abc8",
+  "sv-utils-env": "us.gcr.io/broad-dsde-methods/markw/sv-utils-env:mw-concordance-7251507",
+  "sv_utils_docker": "us.gcr.io/broad-dsde-methods/markw/sv-utils:mw-concordance-7251507",
   "gq_recalibrator_docker": "us.gcr.io/broad-dsde-methods/tbrookin/gatk:0a7e1d86f",
-  "str": "us.gcr.io/broad-dsde-methods/gatk-sv/str:2022-09-15-v0.25.1-beta-b53e58af"
```
Comment (VJalili) [inline inputs/values/dockers.json:34]: I suppose this PR is not related to EH, if so, I wonder how `build_docker.py` ended up incorrectly updating this, while it worked correctly when invoked by the bot.

Comment (mwalker174) [reply inputs/values/dockers.json:34]: Had to do with the way I invoked build_docker.py - my master branch was out of date on the machine where I did the build but my working branch was rebased on the latest master, which contained your recent changes to some str code. I used "master" as the current commit - so it saw the str updates relative to the outdated master and rebuilt the docker.

Comment (VJalili) [reply inputs/values/dockers.json:34]: That explains it, thanks!
